feat!: modular rewrite (#103)

xynydev opened the pull request:
This is an ongoing WIP rewrite for startingpoint to make it more modular. Comments appreciated. Code will be running through shellcheck. 

Made this a 1.0 goal since I wouldn't do a "first release" with a rewrite in the works. Think of the current version as a 0.9 which won't be getting any breaking changes or new features for as long as this is in the works.

Aiming for feature parity but cleaner config files.

Commit:
chore: outline rewrite without code changes
* this does not work yet
* comments & documentation will be added once the rewrite is complete

Commit:
chore: refactor code to support new structure
* this doesn't do much yet and a lot of code was deleted,
functionality will be gradually added back

Commit:
chore: rewriting some parts, new recipe location

Commit:
chore: also copy modules directory into img build

EinoHR commented:
Other things I'll probably do in this PR;
* Move documentation away from the main README and into the new config directory and inside module folders.
* Less inline documentation.
   * It's good _to a certain extent_. I feel that rn it is hurting readability quite a bit, and many of the comments just restate the line without adding context.
   * Example: 
   
       ```
       # Copy public key
       COPY cosign.pub /usr/share/ublue-os/cosign.pub
       ```

EinoHR commented:
I wonder what this error's about. I'll probably remove wallpapers for now.
```
ublue-os-wallpapers-0.1-1.fc38.noarch (@commandline)
Installing: ublue-os-wallpapers-0.1-1.fc38.noarch (local)
ln: failed to create symbolic link '/usr/share/wallpapers/ublue-os': No such file or directory
Building startingpoint from Fedora 38 (ghcr.io/ublue-os/silverblue-main).
Error: 1:1: invalid input text "r"
error building at STEP "RUN rpm-ostree install /tmp/ublue-os-wallpapers-0.1-1.fc38.noarch.rpm &&         chmod +x /tmp/build.sh &&         /tmp/build.sh &&         rm -rf /tmp/* /var/* &&         ostree container commit": error while running runtime: exit status 1
time="2023-08-09T17:41:19Z" level=error msg="exit status 1"
Error: Error: buildah exited with code 1
```

Commit:
chore: remove wallpaper rpm installation (for now)

Commit:
fix: bad yq command syntax
* I'm never trusting AI again!

EinoHR commented:
(the problem was not wallpapers, it was this line, which now is obviously wrong after i read it and not just trust AI blindly)
`TYPE=$(yq r <(echo "$module") '.type')`

Commit:
chore: investigate reading module array

Commit:
fix: make get_yaml_array work
apparently it's not possible to return an array from a function
(eventhough doing so would be nice)

Commit:
chore: launch modules, detect if module inline

Commit:
refactor: remove get yml string, add configdir var

Commit:
feat: running modules from files

Commit:
fix: remove trailing slashes from directory vars

Commit:
fix: bad indentation in common-packages.yml

Commit:
chore: export config directory

Commit:
feat: script runner module

Commit:
refactor: export get_yaml_array

Commit:
Revert "refactor: export get_yaml_array"

This reverts commit c27685fae21ca767c47c63665a03cae55c47fc13.

(get_yaml_array can't be used in modules since it reads from
the $RECIPE_FILE)

Commit:
Merge branch 'template' into modular-rewrite

Commit:
refactor: global get_yaml_array function

Commit:
feat: rpm-ostree module

Commit:
feat: yafti module

Commit:
docs: better inline docs and prints

Commit:
fix: also install and enable yafti in yafti module
oops, forgot this one
there's currently no removing of the firstboot data, though,
but that doesn't affect people much ig

Commit:
chore: export more vars as readonly

Commit:
feat: signing script

Commit:
refactor: rename fedora version to os version
this change makes sense for when in the future it's possible to
build other OS's with this same system
it was done now, because i'm bundling a bunch of breaking changes
into a single update

Commit:
fix: can't set readonly var after declaration
(duh)

Commit:
chore: investigate "no such file or directory"

Commit:
chore: investigate "no such file or directory"

Commit:
fix: give executable permission to scripts

Commit:
fix: don't run arbitrary commands with script
* scripts are now called with their absolute paths instead of ./

xynydev commented on modules/script/script.sh:
```diff
@@ -3,13 +3,13 @@
-for CMD in "${RUN[@]}"; do
-    echo "Running command: $CMD"
-    eval "$CMD"
+for SCRIPT in "${SCRIPTS[@]}"; do
```
this thing just is not working????? 
no matter how I put it, it [can't find the script files](https://github.com/ublue-os/startingpoint/actions/runs/5848195554/job/15855225241) when trying to run them, but [can find them with `ls`](https://github.com/ublue-os/startingpoint/actions/runs/5848141493/job/15855113964).

time to take a break...

Commit:
docs: move & start rewriting configuration docs

Commit:
chore: remove changelog
* this was added earlier when making the ISO pr,
and it is not going to be updated anymore

Commit:
docs: a ginormous amount of docs for the modules

Commit:
chore!: remove autorun.sh
i think the existence of multiple built-in ways to add scripts
makes it more complicated,
and paradoxically, autorun was made to uncomplicate things

with the removal of clear "pre" and "post" hooks,
only a linear list of modules, the filesystem-based method
is ever more complicated

if such a feature is ever wanted again,
it should be implemented inside the script module

Commit:
chore: bring recipe.yml to a default state

Commit:
chore: remove outdated script-related docs

Commit:
chore: clean up bling a bit

Commit:
feat: ublue-os/bling module
this makes all bling components optional,
while making more of them accessible

Commit:
chore: clean build.yml, make matrix more visible
there were huge blocks of good documentation,
but it was in the way of easy access to the matrix section,
 which ordinary users have to update

 it was also only about build triggers, which is a pretty clear topic
 much more valuable would be explaining each build step

 i condesensed the docs into a smaller section atop
 with only generally useful information

Commit:
chore: remove empty usr/bin dir

Commit:
docs: expand main customization section

Commit:
chore!: remove setup-flatpaks
* it doesn't make sense with the module system anymore

Commit:
docs: rewrite the just section

Commit:
chore: re-add branch selection for push
* same-repo PRs build twice

Commit:
fix: newlines in rpm-ostree cmds errors
i'm not sure if this approach is better than iterating over the array
but it should now work in the same way it did previously

xynydev commented on modules/rpm-ostree/rpm-ostree.sh:
```diff
@@ -15,13 +15,15 @@ fi
     echo "Installing RPMs"
-    echo "Installing: ${INSTALL[*]}"
-    rpm-ostree install "${INSTALL[@]}"
+    INSTALL_STR=$(echo "${INSTALL[*]}" | tr -d '\n')
```
I don't get this error (which this commit didn't fix). Running `rpm-ostree install "python3-pip" "libadwaita"` works, but running `rpm-ostree install $INSTALL_STR` where `$INSTALL_STR` is `"python3-pip" "libadwaita"` doesn't. 

Iterating the array could fix it, but this approach did work in [the previous versions](https://github.com/ublue-os/startingpoint/blob/d04ad0f78c2b3a7c32cc4c75bafff0f267535af4/scripts/build.sh#L59C29-L59C29) where the yaml array was output as yaml and not json.

Or maybe it was directly a shell array hmm, let me...

xynydev replied:
That fixed it, now there's another error...

Commit:
fix: don't output yaml array as json

Commit:
chore: revert
e27eb25 fix

Commit:
docs: document added/removed packages

EinoHR commented:
```
Resolving dependencies...done
error: Could not depsolve transaction; 2 problems detected:
Problem 1: conflicting requests
Problem 2: conflicting requests
```

Working all day on this, now at a dead end. Need some food and a two hour walk. The script.sh error is probably still around, just haven't gotten to it yet with these rpm-ostree errors appearing.

This PR is by no means ready, but the content is finished, so please feel free to look around.

If someone could look at this rpm-ostree conflicting request problem, that'd be great.

Commit:
chore: rm ublue-os/startingpoint in iso docs
* automatic setup replaces it with the repository's details,
which makes it confusing to read

Commit:
chore: use same package inst/rm logic as upstream
upstream referring to build.sh in main and bluefin

xynydev committed:
Merge branch 'template' into modular-rewrite

Commit:
chore: try removing pkgs to remove?

Commit:
Merge branch 'modular-rewrite' of github.com:ublue-os/startingpoint into modular-rewrite

Commit:
fix: put .sh suffix outside of quotes?

xynydev commented on modules/bling/bling.sh:
```diff
@@ -13,5 +13,5 @@ find "$PWD" -type f -exec chmod +x {} \;
 for ITEM in "${INSTALL[@]}"; do
     echo "Pulling from bling: $ITEM"
-    eval "$PWD/$ITEM.sh"
+    eval "$PWD/$ITEM".sh
```
I don't get this error, the `.sh` is clearly there but it's not in the error. I tried both the old and new snippet locally and both worked..

xynydev replied:
From the empty space,, it could be that $ITEM has a newline
```
=== Launching module of type: bling ===
Pulling from bling: fonts

/tmp/modules/bling/bling.sh: line 16: /tmp/modules/bling/installers/fonts: No such file or directory
```

xynydev replied:
That was it. Seems that all variables in all arrays contain a trailing newline, but that doesn't cause big issues anywhere else.

Commit:
fix: remove trailing newline from $ITEM
using a clever bash-ism

Commit:
docs: some more inline docs for modules
(only confusing parts)

Commit:
chore: try removing something other than ff
removing firefox gives some cryptic errors,
let's try removing something else
to investigate

Commit:
fix: rm newlines from rpm-ostree cmds
basically re-adds the e27eb25 changes
which were later reverted
in favor of the way packages.json apparently worked

Commit:
chore: revert recipe to default configuration

Commit:
fix: ? remove quotes, add wordsplitting

Commit:
chore: back to array syntax in rpm-ostree

EinoHR commented:
Ok with [ac8be6e](https://github.com/ublue-os/startingpoint/pull/135/commits/ac8be6e92bcef06efc48274c84ae2755b2cb100c) we're back at the error while [4ce8e91](https://github.com/ublue-os/startingpoint/pull/135/commits/4ce8e919b9d7186738a96d06470d43c6fe93613a) still builds perfectly fine. So I guess I'll just revert the new commit, let shellcheck be mad at me and go on with my day...

Commit:
chore: Revert "chore: back to array syntax in rpm-ostree"

This reverts commit ac8be6e92bcef06efc48274c84ae2755b2cb100c.

gerblesh commented on modules/bling/installers/ublue-update.sh:
```diff
@@ -0,0 +1,6 @@
+
+# Tell build process to exit if there are any errors.
+set -oue pipefail
+
```
may want to check if `ublue-os-update-services` is in the installed system and remove it, as it conflicts with ublue-update:
```bash
# Check if ublue-os-update-services rpm is installed, these services conflict with ublue-update
if rpm -q ublue-os-update-services > /dev/null; then
    rpm-ostree override remove ublue-os-update-services
fi
```

xynydev replied:
[2208ea8](https://github.com/ublue-os/startingpoint/pull/135/commits/2208ea88e50bb504c440783912dd0cd56f62d008)

Commit:
fix: remove ub update services in ub updater inst

xynydev commented on modules/rpm-ostree/README.md:
```diff
@@ -0,0 +1,22 @@
+
+Then the module installs the packages declared under `install:` using `rpm-ostree install`, and lastly, it removes the packages declared under `remove:` using `rpm-ostree override remove`.
+
+Unfortunately, currently `rpm-ostree override remove`, and this module, might not be able to remove packages installed in image builds. Packages included by Fedora, such as Firefox can still be removed, though.
```
Is this accurate? Every time I wonder this out loud I get a bit of a different answer. Worth including in the docs, even? What about `remove` being like `git rm` in that it doesn't reclaim disc space?

gerblesh replied:
I think it is able to remove packages installed in image builds iirc, I'll need to check with that but i'm pretty positive that is an option

Commit:
feat: integrate optfix into rpm-ostree module
Co-Authored By: @AuraHerreroRuiz
bulk of the code from:
github.com/AuraHerreroRuiz/auraos/
blob/f83e90ea9f1dddc50630f4634d835f085adf4a99/scripts/pre/optpackages.sh

EinoHR commented:
With the PR nearing completion, we should start somewhat extensive 
### Testing
#### Everyone should/could test
* Creating a repo from the `modular-rewrite` branch, reading docs, applying customizations, building an image, rebasing to it (vm or metal, shouldn't make a difference)
* Trying to merge changes from the new branch to see how it works out and how much work needs to be redone.
#### I'd rather have someone else test
* Installing packages that install into `/opt`
* Figuring out ways to break / misunderstand things
* Reading docs for weird sentences and incorrect "facts"

Commit:
docs: fix: it's not dnkmmr it's determinate

xynydev commented:
Ok, doing some testing in [this repo](https://github.com/SchemingSandy/vauxirrific) (as always). Got an error with the repo DL stuff and will be investigating. https://github.com/SchemingSandy/vauxirrific/actions/runs/6059363398/job/16442414493

Commit:
fix: remove newlines before wget
including the newlines causes filenames to be postfixed with %0A
and thus the repo files to not work

this commit uses the same bashism as above line to remove newlines
from the $REPO var before using it in the wget command

gerblesh commented:
it when you do --install it seems to only expect a single package as specified in the help message:
```
Application Options:
  --install=PKG           Overlay additional package
```

xynydev commented:
> when you use the flag --install rpm-ostree seems to only expect a single package as specified in the help message:
> 
> ```
> Application Options:
>   --install=PKG           Overlay additional package
> ```

The way I use `--install` is directly copied from ubluemain https://github.com/ublue-os/main/blob/1cfad9e1f728707f65864125959126dadfe3ab29/main-install.sh#L55, that shouldn't be the problem here (as evidenced by the build working when not trying to remove `vim`, the current errors have been ghcr perms issues).

Commit:
fix: bling justfile installer nested just dir
idon't create another just directory inside the basic just directory,
just copy the contents

Commit:
fix: some more bling installer file copy issues
(brainfarts), the nix install one is just...

Commit:
fix: move globs outside of ""

gerblesh committed:
feat: added systemd and files module (#142)

* feat: add files and systemd module

* fix: yaml formatting

* fix: yaml formatting

* fix: remove comment completely

* fix: yaml formatting

* docs: add back inline comment

* reformat: rename variables

* fix: fix systemd escaped string

* fix: fix systemd service formatting with printf

* fix: attempting to fix systemd module problems

* chore: remove debug config and code from systemd module

* docs: added WIP docs for systemd, reworked files README

* docs: added more detail for systemd module

* docs: update READMEs to be more consistent

* docs: remove unneeded sentence

* docs: remove unneeded sentence

* chore: fix issues described in PR review

* docs: fix markdown formatting

* docs: fix markdown formatting

* docs: better markdown

xynydev commented:
Thanks to @gerblesh  for the `systemd` and `files` modules. More testing to be done this week, and reading the docs through and through.

xynydev committed:
Merge branch 'template' into modular-rewrite

gerblesh committed:
fix: removed issues in rpm-ostree module and ublue-update bling installer  (#144)

* fix(rpm-ostree): format string in optfix
This fixes optfix and makes the symlink not broken

* fix: change rpm-ostreed auto update policy in ublue-update installer

Commit:
fix: add .sh suffix for dconf update service inst

Commit:
docs: some docs fixes, better formatting, etc

Commit:
feat: yafti module installs yafti deps

Commit:
docs: missing space

Commit:
feat: enable dconf update service in installer

Commit:
docs: mention depsolve & more info in walkthrough

Commit:
chore: don't use _STR like array

xynydev commented:
I've done some testing [here](https://github.com/SchemingSandy/vauxirrific/actions) and read through the docs. Everything looks good, but still can't remove `vim` due to `depsolve` issues. I'll just leave a mention of it in the docs and go on with my life.

xynydev commented on modules/rpm-ostree/README.md:
````diff
@@ -21,3 +21,13 @@ remove:
 ```
+
+
+## Known issues
````
.

Commit:
docs: add notice of rewrite

Commit:
docs: point reader toward discovering new features

tulilirockz commented:
Hello Xyny! Remember when the bling repo was made to dedupe some aspects in Startingpoint? Well, this PR makes it so every repo that uses the new Startingpoint project needs to have all it's modules inside of it. Maybe we could move them to the bling repo and clone them to the project. That would make it so no one needs to have the standard modules in their repos, and would also enable people to make repos just for their custom modules, e.g. someone could make a "ublue-unofficial-extensions" repo or something like that!

Summarizing: Instead of having `COPY modules /tmp/modules/` we could have `COPY --from=ghcr.io/ublue-os/bling:latest /startingpoint-std-modules(something like that) /tmp/modules`

xynydev commented:
@tulilirockz I love that idea. I'll drop a bling PR, and make some modifications to the website PR to pull in the docs. I sent some more cluttered thoughts over on discord, but I think this is overall beneficial. The size of startingpoint would actually *shrink*.

Commit:
refactor: fetch modules from bling

tulilirockz commented:
> I wonder what this error's about. I'll probably remove wallpapers for now.
> 
> ```
> ublue-os-wallpapers-0.1-1.fc38.noarch (@commandline)
> Installing: ublue-os-wallpapers-0.1-1.fc38.noarch (local)
> ln: failed to create symbolic link '/usr/share/wallpapers/ublue-os': No such file or directory
> Building startingpoint from Fedora 38 (ghcr.io/ublue-os/silverblue-main).
> Error: 1:1: invalid input text "r"
> error building at STEP "RUN rpm-ostree install /tmp/ublue-os-wallpapers-0.1-1.fc38.noarch.rpm &&         chmod +x /tmp/build.sh &&         /tmp/build.sh &&         rm -rf /tmp/* /var/* &&         ostree container commit": error while running runtime: exit status 1
> time="2023-08-09T17:41:19Z" level=error msg="exit status 1"
> Error: Error: buildah exited with code 1
> ```

Also, this most likely is fixed now :) I just made it sure the /usr/share/wallpapers/ublue-os folder and it's parents get created in the rpm post-installation stage, check it out!
